test/hotspot/jtreg/compiler/lib/verify/Verify.java

```diff
@@ -202,83 +310,200 @@ private static void checkEQimpl(MemorySegment a, MemorySegment b, String context
      */
-    private static void checkEQimpl(Object[] a, Object[] b, String context) {
+    private void checkEQimpl(Object[] a, Object[] b, String field, Object aParent, Object bParent) {
         // (1) Length mismatch
         if (a.length != b.length) {
             System.err.println("ERROR: Verify.checkEQ failed: length mismatch: " + a.length + " vs " + b.length);
+            print(a, b, field, aParent, bParent);
             throw new VerifyException("Object array length mismatch.");
         }
 
         for (int i = 0; i < a.length; i++) {
             // Recursive checkEQ call.
-            checkEQ(a[i], b[i], "[" + i + "]" + context);
+            checkEQdispatch(a[i], b[i], "[" + i + "]", a, b);
         }
     }
 
-    private static void print(Object a, String context) {
-        if (a == null) {
-            System.err.println("  " + context + ": null");
+    private void checkEQArbitraryClasses(Object a, Object b) {
+        Class c = a.getClass();
+        while (c != Object.class) {
+            for (Field field : c.getDeclaredFields()) {
+                Object va = null;
+                Object vb = null;
+                try {
+                    field.setAccessible(true);
+                    va = field.get(a);
+                    vb = field.get(b);
+                } catch (IllegalAccessException e) {
+                    throw new VerifyException("Failure to access field: " + field + " of " + a);
+                }
+                checkEQdispatch(va, vb, field.getName(), a, b);
+            }
+            c = c.getSuperclass();
+        }
+    }
+
+    private void print(Object a, Object b, String field, Object aParent, Object bParent) {
+        System.err.println("  aParent: " + aParent);
+        System.err.println("  bParent: " + bParent);
+        System.err.println("  field:   " + field);
+        System.err.println("  a:       " + a);
+        System.err.println("  b:       " + b);
+    }
+
+    private boolean checkAlreadyVisited(Object a, Object b, String field, Object aParent, Object bParent) {
+        // Boxed primitives are not guaranteed to be the same Object for the same primitive value.
+        // Hence, we cannot use the mapping below. We test these boxed primitive types by value anyway,
+        // and they are no recursive structures, so there is no point in optimizing here anyway.
+        switch(a) {
+            case Boolean x -> { return false; }
+            case Byte x -> { return false; }
+            case Short x -> { return false; }
+            case Character x -> { return false; }
+            case Integer x -> { return false; }
+            case Long x -> { return false; }
+            case Float x -> { return false; }
+            case Double x -> { return false; }
+            default -> {}
+        }
+
+        Object bPrevious = a2b.get(a);
+        Object aPrevious = b2a.get(b);
+        if (aPrevious == null && bPrevious == null) {
+            // Record for next time.
```

Comment:
Can you explain, maybe as comment at `checkAlreadyVisited()`, why we want to have these caches?

Reply:
Added documentation :)